### ISSUE TITLE
Integration errors with certain n(z)

### ISSUE THREAD

**Issue by tilmantroester** [LSSTDESC/CCL#927]
CCL fails to create (weak lensing) tracers for certain n(z). A minimal example is given below. Enabling debug mode shows that the issue appears when trying to normalise the n(z). Curiously, there should be an error message being set (https://github.com/LSSTDESC/CCL/blob/master/src/ccl_tracers.c#L85) but for some reason that gets lost somewhere.

```python
import numpy as np
import pyccl as ccl

z, nz = np.loadtxt("bad_nz.txt", unpack=True)

ccl.debug_mode(True)
ccl_cosmo = ccl.CosmologyVanillaLCDM()
ccl_tracer = ccl.WeakLensingTracer(ccl_cosmo, (z, nz))
```
[bad_nz.txt](https://github.com/LSSTDESC/CCL/files/8489362/bad_nz.txt)

**Comment by nikfilippas** [comment]
When I tried to reproduce this some days ago, adding `cosmo` as an extra argument in the checks of `tracers.py` brings the erro message back. Maybe it's worth adding those to your PR as well.

**Comment by tilmantroester** [comment]
Further investigation showed that the fixed order quadrature used in the tracer creation in CCL doesn't cope with this n(z). Doing the normalisation integral in python independently of CCL shows that fixed order quadrature fails miserably and adaptive quadrature fails to converge as well with default settings: 

```python
import scipy.integrate
import scipy.interpolate

z, n = np.loadtxt("bad_nz.txt", unpack=True)
nofz = scipy.interpolate.InterpolatedUnivariateSpline(z, n)

norm = {}
norm["trapz"] = np.trapz(n, z)
norm["spline_integral"] = nofz.integral(z[0], z[-1])
norm["adaptive_quad"] = scipy.integrate.quad(nofz, z[0], z[-1])[0]
for i in range(1, 8):
    norm[f"fixed_quad_order_{i}"] = scipy.integrate.fixed_quad(nofz, z[0], z[-1], n=i)[0]

for name, value in norm.items():
    print(f"{name:<20}: {value}")
```
```
trapz               : 1.0
spline_integral     : 1.000000000000001
adaptive_quad       : 0.9999361333214825
fixed_quad_order_1  : 0.08137690272918875
fixed_quad_order_2  : 3.7549939231261926
fixed_quad_order_3  : 0.03843028804193624
fixed_quad_order_4  : 0.39222769522755374
fixed_quad_order_5  : 1.371894742387018
fixed_quad_order_6  : 1.6210543006866418
fixed_quad_order_7  : 0.42355914676853335
```
For the n(z) normalisation, spline integration is probably useful, since the spline of the integrand already exists. But that still causes issues in `ccl_get_lensing_mag_kernel`. Integration failure there just doesn't cause an error, just a warning, which by default is silent.

**Comment by damonge** [comment]
OK, I agree we should move this to spline integration (or at least to make it default to that).
I've been annoyed by this in the past.

**Comment by damonge** [comment]
BTW, is this a problem with the normalisation of the N(z) or with the integral needed to get the lensing kernel?

**Comment by tilmantroester** [comment]
Spline integration fixes the normalisation but then the integral for the lensing kernel fails. We've just not been checking that second integral so far.

**Comment by tilmantroester** [comment]
I found that the errors in the lensing integral are due to round-off errors that happen in the high-z tails of the n(z), even when smoothing the noisy n(z) provided above. Now, in practice this shouldn't matter too much, since the value of the lensing kernel at high redshift is small compared to its peak.
![lensing_integrand](https://user-images.githubusercontent.com/26582252/164736390-214a240f-c44d-41c8-8659-89e8845aa333.png)

So a solution would be to check if the integration error is small compared to the value at the ~peak of the kernel. In other words, add an absolute error tolerance instead of just a relative error tolerance.

**Comment by damonge** [comment]
this solution sounds good to me

**Comment by tilmantroester** [comment]
Addressed with #928
